File: Hard/Burning Tree/burning-tree.py
#User function Template for python3

class Solution:
    def buildmap(self,node,parent_val,graph):# form undirected graph frm tree
        if not node:
            return
        if node.data not in graph:
            graph[node.data]=[]
            if(parent_val is not None):
                graph[parent_val].append(node.data)
                graph[node.data].append(parent_val)
            self.buildmap(node.left,node.data,graph)
            self.buildmap(node.right,node.data,graph)
    def minTime(self, root,target):
        graph={}
        self.buildmap(root,None,graph)
        # for node,neighbors in graph.items():
        #     print(node,neighbors)
        # print(graph)
        q=[(target)]
        vis=set([target])
        t=0
        while(q):
            size=len(q)
            for i in range(size):
                node=q.pop(0)
                for neighbor in graph[node]:
                    if neighbor in vis:
                        continue
                    vis.add(neighbor)
                    q.append(neighbor)
            t+=1
        return t-1
        

#{ 
 # Driver Code Starts
#Initial Template for Python 3

from collections import deque

# Tree Node
class Node:
    def __init__(self, val):
        self.right = None
        self.data = val
        self.left = None

# Function to Build Tree   
def buildTree(s):
    #Corner Case
    if(len(s)==0 or s[0]=="N"):           
        return None
        
    # Creating list of strings from input 
    # string after spliting by space
    ip=list(map(str,s.split()))
    
    # Create the root of the tree
    root=Node(int(ip[0]))                     
    size=0
    q=deque()
    
    # Push the root to the queue
    q.append(root)                            
    size=size+1 
    
    # Starting from the second element
    i=1                                       
    while(size>0 and i<len(ip)):
        # Get and remove the front of the queue
        currNode=q[0]
        q.popleft()
        size=size-1
        
        # Get the current node's value from the string
        currVal=ip[i]
        
        # If the left child is not null
        if(currVal!="N"):
            
            # Create the left child for the current node
            currNode.left=Node(int(currVal))
            
            # Push it to the queue
            q.append(currNode.left)
            size=size+1
        # For the right child
        i=i+1
        if(i>=len(ip)):
            break
        currVal=ip[i]
        
        # If the right child is not null
        if(currVal!="N"):
            
            # Create the right child for the current node
            currNode.right=Node(int(currVal))
            
            # Push it to the queue
            q.append(currNode.right)
            size=size+1
        i=i+1
    return root

if __name__=="__main__":
    t=int(input())
    for _ in range(t):
        line=input()
        target=int(input())
        root=buildTree(line)
        print(Solution().minTime(root,target))

# } Driver Code Ends

File: Difficulty: Hard/Burning Tree/burning-tree.cpp
class Solution {
public:
    int findMaxDist(map<Node*, Node*>& m, Node* target) {
        queue<Node*> q;
        q.push(target);
        map<Node*, int> vis;
        vis[target] = 1;
        int maxi = 0;
        while (!q.empty()) {
            int len = q.size();
            int flag = 0;
            for (int i = 0; i < len; i++) {
                auto node = q.front();
                q.pop();
                if (node->left && !vis[node->left]) {
                    flag = 1;
                    vis[node->left] = 1;
                    q.push(node->left);
                }
                if (node->right && !vis[node->right]) {
                    flag = 1;
                    vis[node->right] = 1;
                    q.push(node->right);
                }
                if (m[node] && !vis[m[node]]) {
                    flag = 1;
                    vis[m[node]] = 1;
                    q.push(m[node]);
                }
            }
            if (flag) maxi++;
        }
        return maxi;
    }

    Node* bfstoMapParents(Node* root, map<Node*, Node*>& m, int start) {
        queue<Node*> q;
        q.push(root);
        Node* res = NULL;
        while (!q.empty()) {
            Node* node = q.front();
            if (node->data == start){
                res=node;
            }
            q.pop();
            if (node->left) {
                m[node->left] = node;
                q.push(node->left);
            }
            if (node->right) {
                m[node->right] = node;
                q.push(node->right);
            }
        }
        return res;
    }

    int minTime(Node* root, int target) {
        map<Node*, Node*> m;
        Node* start = bfstoMapParents(root, m, target);
        int maxi = findMaxDist(m,start);
        return maxi;
    }
};



